chain/epoch_manager/src/shard_assignment.rs

```diff
@@ -103,19 +106,7 @@ fn assign_with_possible_repeats<T: HasStake + Eq, I: Iterator<Item = T>>(
                     // one to any shard and move to next cp.
                     break;
                 }
-                Some(top) if result[usize::try_from(top.2).unwrap()].contains(&cp) => {
-                    // This chunk producer is already assigned to this shard.
-                    // Pop the shard from the heap for now and try assigning the
-                    // producer to the next shard.  (We’ll look back at the
-                    // shard once we figure out what to do with current `cp`).
-                    //
-                    // TODO(mina86): The contains check in the condition makes
-                    // this an O(N^2) algorithm.  At the moment there aren’t too
-                    // many chunk producers so it should be fine but if that’s
-                    // becomes an issue we should switch to a hash set.
-                    buffer.push(PeekMut::pop(top));
-                }
-                Some(mut top) => {
+                Some(mut top) if seen.insert((top.2, cp_index)) => {
```

Comment:
I'm missing something.

What's preventing us from adding the same cp twice to the same shard?

as cp_iter is 'cycle' - it means that it can have the same cp multiple times with multiple 'cp_indexes', right?

Reply:
The cycle is done after enumeration so the same cp will have the same cp_index.

Reply:
ah, ok - that's what I've missed. Let's comment about it then (that cp_index is basically like a 'hash'/unique_identifier of the cp) and we're good to go.